core/modal.c

```diff
@@ -0,0 +1,371 @@
+/*************
+Copyright (c) 2021, Kiel University.
+
+Redistribution and use in source and binary forms, with or without modification,
+are permitted provided that the following conditions are met:
+
+1. Redistributions of source code must retain the above copyright notice,
+   this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright notice,
+   this list of conditions and the following disclaimer in the documentation
+   and/or other materials provided with the distribution.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY
+EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+***************/
+
+/**
+ * Runtime infrastructure for modes in the C target of Lingua Franca.
+ * This file contains functions that handle the processing of mode transitions.
+ * It works together with other changes across the runtime implementation.
+ *
+ * Any mode related code will only work in the presence of the MODAL_REACTORS
+ * definition.
+ *
+ * This file is intended for direct include in reactor_common.c.
+ *
+ * @author{Alexander Schulz-Rosengarten <als@informatik.uni-kiel.de>}
+ * @author{Soroush Bateni <soroush@utdallas.edu}
+ */
+#ifdef MODAL_REACTORS
+
+#include "reactor.h"
+
+// Forward declaration of functions and variables supplied by reactor_common.c
+void _lf_trigger_reaction(reaction_t* reaction, int worker_number);
+event_t* _lf_create_dummy_events(trigger_t* trigger, instant_t time, event_t* next, microstep_t offset);
+extern reaction_t** _lf_startup_reactions;
+extern int _lf_startup_reactions_size;
+extern pqueue_t* event_q;
+
+/**
+ * Checks whether the given mode is currently considered active.
+ * This includes checking all enclosing modes.
+ * If any of those is inactive, then so is this one.
```

Comment:
This doc is confusing. It implies that every enclosed mode within an active mode is active. But shouldn't only one of those be active?

Also, I would change to use imperative case. For example, the doc could be:
```
 * Return true if the given mode is active.
 * A given mode is active if it is marked active and all its enclosing modes are active.
```

Reply:
I think the comment is correct, since it is talking about enclosing modes (up in the hierarchy), not enclosed modes (down in the hierarchy).